examples/PlaneCurve.jl

```diff
@@ -0,0 +1,273 @@
+  elseif typeof(C) == ProjPlaneCurve{typeof(lc(C.eq))}
+    D = factor(C.eq.f)
+    C.components = Dict(ProjPlaneCurve(x) => D.fac[x] for x in keys(D.fac))
+  end
+end
+
+################################################################################
+# Components of the curve
+
+@doc Markdown.doc"""
+    curve_components(C::PlaneCurve)
+
+Return a dictionary containing the irreducible components of `C` and their multiplicity.
+"""
+function curve_components(C::PlaneCurve)
+  if isempty(C.components)
+    compo(C)
+  end
+  return C.components
+end
+
+################################################################################
+# Check irreducibility.
+# TODO: change for a direct irreducibility check when available.
+
+@doc Markdown.doc"""
+    isirreducible(C::PlaneCurve)
+
+Return `true` if `C` is irreducible, and `false` otherwise.
+"""
+function Oscar.isirreducible(C::PlaneCurve)
+  if isempty(C.components)
+     compo(C)
+  end
+  if length(C.components) != 1
+     return false
+  else
+     return all(isone, values(C.components))
+  end
+end
+
+################################################################################
+# Check reducedness by computing a factorization
+# TODO: change for a direct squarefree check when available.
+
+@doc Markdown.doc"""
+    isreduced(C::PlaneCurve)
+
+Return `true` if `C` is reduced, and `false` otherwise.
+"""
+function Oscar.isreduced(C::PlaneCurve)
+  if isempty(C.components)
+    compo(C)
+  end
+  return all(isone, values(C.components))
+end
+
+################################################################################
+# Compute squarefree equation.
+# TODO: change for a direct squarefree computation when available.
+
+@doc Markdown.doc"""
+    reduction(C::PlaneCurve)
+
+Return the plane curve defined by the squarefree part of the equation of `C`.
+"""
+function reduction(C::PlaneCurve)
+  if isempty(C.components)
+     compo(C)
+  end
+  F = prod(D -> D.eq, keys(C.components))
+  if typeof(C) == AffinePlaneCurve{typeof(lc(C.eq))}
+    rC = AffinePlaneCurve(F)
+    rC.components = Dict(AffinePlaneCurve(D.eq) => 1 for D in keys(C.components))
+  elseif typeof(C) == ProjPlaneCurve{typeof(lc(C.eq))}
+    rC = ProjPlaneCurve(F)
+    rC.components = Dict(ProjPlaneCurve(D.eq) => 1 for D in keys(C.components))
+  end
+  return rC
+end
+
+################################################################################
+# Union of two plane curves of the same type (with multiplicity)
+
+@doc Markdown.doc"""
+   union(C::PlaneCurve{S}, D::PlaneCurve{S}) where S <: FieldElem
+
+Return the union of `C` and `D` (with multiplicity).
+"""
+function Base.union(C::PlaneCurve, D::PlaneCurve)
+  typeof(C) == typeof(D) || error("The curves do not have the same type.")
+  F = C.eq*D.eq
+  if typeof(C) == ProjPlaneCurve{typeof(lc(C.eq))}
+     return ProjPlaneCurve(F)
+  elseif typeof(C) == AffinePlaneCurve{typeof(lc(C.eq))}
+     return AffinePlaneCurve(F)
+  end
+end

```

Comment:
Could also do this:
```suggestion
function Base.union(C::T, D::T)  where T <: PlaneCurve
     return T(C.eq*D.eq)
end
```

Reply:
I am working on a fix for some type errors, I allowed decorated or non decorated polynomial as an input for ProjPlaneCurve for the current working version, but it leads to some polynomials not leaving in the same ring. I am writing now some functions for decorated polynomials to stay type consistent.
To @fingolfin: as I already answered to @thofma  in another question, T(C.eq) does not work, it throws me an error.

Reply:
I think the problem here is that `T` is not equal to `AffinePlaneCurve` or `ProjPlaneCurve` but rather to a concrete parametrization of those, i.e. `ProjPlaneCurve{S}` for a specific type `S`/ And the type of `C.eq*D.eq` does not match.

I think that hints at a (hopefully minor) design issue -- but I am sure it could be resolved; it doesn't have to be right now, though. In the meantime, this ought to work (or just leave this bit as it is, and we deal with it later -- perhaps @thofma or @fieker or me can have a look at your code once it is merged to figure this out):
```suggestion
function Base.union(C::T, D::T)  where T <: AffinePlaneCurve
     return AffinePlaneCurve(C.eq*D.eq)
end
function Base.union(C::T, D::T)  where T <: ProjPlaneCurve
     return ProjPlaneCurve(C.eq*D.eq)
end
```